src/analysis/seminorm.lean

```diff
@@ -1212,4 +1249,287 @@ end
+  add_group_filter_basis_of_comm (seminorm_basis_zero p)
+  (seminorm_basis_zero_nonempty p)
+  (seminorm_basis_zero_intersect p)
+  (seminorm_basis_zero_zero p)
+  (seminorm_basis_zero_add p)
+  (seminorm_basis_zero_neg p)
+
+lemma seminorm_basis_zero_smul (p : ι → seminorm 𝕜 E) (U) (hU : U ∈ seminorm_basis_zero p) :
+  (∃ (V : set 𝕜) (H : V ∈ 𝓝 (0 : 𝕜)) (W : set E)
+  (H : W ∈ (seminorm_add_group_filter_basis p).sets), V • W ⊆ U) :=
+begin
+  rcases (seminorm_basis_zero_iff p U).mp hU with ⟨ι', r, hr, hU⟩,
+  refine ⟨metric.ball 0 r.sqrt, metric.ball_mem_nhds 0 (real.sqrt_pos.mpr hr), _⟩,
+  refine ⟨(ι'.sup p).ball 0 r.sqrt, seminorm_basis_zero_mem p ι' (real.sqrt_pos.mpr hr), _⟩,
+  refine set.subset.trans (ball_smul_ball (ι'.sup p) r.sqrt r.sqrt) _,
+  rw [hU, real.mul_self_sqrt (le_of_lt hr)],
+end
+
+lemma seminorm_basis_zero_smul_left (p : ι → seminorm 𝕜 E) (x : 𝕜) (U : set E)
+  (hU : U ∈ seminorm_basis_zero p) : (∃ (V : set E)
+  (H : V ∈ (seminorm_add_group_filter_basis p).sets), V ⊆ (λ (y : E), x • y) ⁻¹' U) :=
+begin
+  rcases (seminorm_basis_zero_iff p U).mp hU with ⟨ι', r, hr, hU⟩,
+  rw hU,
+  by_cases h : x ≠ 0,
+  { rw [(ι'.sup p).smul_ball_preimage 0 r x h, smul_zero],
+    use (ι'.sup p).ball 0 (r / ∥x∥),
+    exact ⟨seminorm_basis_zero_mem p ι' (div_pos hr (norm_pos_iff.mpr h)), subset.rfl⟩ },
+  refine ⟨(ι'.sup p).ball 0 r, seminorm_basis_zero_mem p ι' hr, _⟩,
+  simp only [not_ne_iff.mp h, subset_def, mem_ball_zero, hr, mem_univ, seminorm.zero,
+    implies_true_iff, preimage_const_of_mem, zero_smul],
+end
+
+lemma seminorm_basis_zero_smul_right (p : ι → seminorm 𝕜 E) (v : E) (U : set E)
+  (hU : U ∈ seminorm_basis_zero p) : (∀ᶠ (x : 𝕜) in 𝓝 0, x • v ∈ U) :=
+begin
+  rcases (seminorm_basis_zero_iff p U).mp hU with ⟨ι', r, hr, hU⟩,
+  rw [hU, filter.eventually_iff],
+  simp_rw [(ι'.sup p).mem_ball_zero, (ι'.sup p).smul],
+  by_cases h : 0 < (ι'.sup p) v,
+  { simp_rw (lt_div_iff h).symm,
+    rw ←_root_.ball_zero_eq,
+    exact metric.ball_mem_nhds 0 (div_pos hr h) },
+  simp_rw [le_antisymm (not_lt.mp h) ((ι'.sup p).nonneg v), mul_zero, hr],
+  exact is_open.mem_nhds is_open_univ (mem_univ 0),
+end
+
+/-- The `module_filter_basis` induced by the filter basis `seminorm_basis_zero`. -/
+def seminorm_module_filter_basis (p : ι → seminorm 𝕜 E) : module_filter_basis 𝕜 E :=
+{ to_add_group_filter_basis := seminorm_add_group_filter_basis p,
+  smul' := seminorm_basis_zero_smul p,
+  smul_left' := seminorm_basis_zero_smul_left p,
+  smul_right' := seminorm_basis_zero_smul_right p }
+
+end filter_basis
+
+section bounded
+
+variables [normed_field 𝕜] [add_comm_group E] [module 𝕜 E] [add_comm_group F] [module 𝕜 F]
+
+/-- The proposition that a linear map is bounded between spaces with families of seminorms. -/
+def is_bounded (p : ι → seminorm 𝕜 E) (q : ι' → seminorm 𝕜 F) (f : E →ₗ[𝕜] F) : Prop :=
+  ∀ i : ι', ∃ s : finset ι, ∃ C : ℝ≥0, C ≠ 0 ∧ (q i).comp f ≤ C • s.sup p
+
+lemma is_bounded_singleton (ι' : Type*) [inhabited ι']
```

Comment:
Why `singleton`? Wouldn't it be better to call it `const`?

Reply:
you are right, I named it singleton, because I of the `finset` singleton stuff, but it is a constant map.